tests/unit/lib/secrets/vault_csi_tests.py

```diff
@@ -0,0 +1,271 @@
+            "request_id": "8487d906-2154-0151-d07e-57f41447326a",
+            "lease_id": "",
+            "lease_duration": 2764800,
+            "renewable": False,
+            "data": {
+                "password": "new_password",
+                "type": "credential",
+                "username": "new_reddit",
+            },
+            "warnings": None,
+        },
+    }
+)
+
+
+class StoreTests(unittest.TestCase):
+    def setUp(self):
+        self.csi_dir = new_fake_csi(EXAMPLE_SECRETS_DATA)
+
+    def tearDown(self):
+        shutil.rmtree(self.csi_dir)
+
+    def test_can_load_credential_secret(self):
+        secrets_store = get_secrets_store(str(self.csi_dir))
+        data = secrets_store.get_credentials("secret/example-service/example-secret")
+        assert data.username == "reddit"
+        assert data.password == "password"
+
+    def test_get_raw_secret(self):
+        secrets_store = get_secrets_store(str(self.csi_dir))
+        data = secrets_store.get_raw("secret/example-service/nested/example-nested-secret")
+        assert data == {"password": "password", "type": "credential", "username": "reddit"}
+
+    def test_get_simple_secret(self):
+        secrets_store = get_secrets_store(str(self.csi_dir))
+        data = secrets_store.get_simple("secret/simple-secret")
+        assert data == b"simply a secret"
+
+    def test_get_versioned_secret(self):
+        secrets_store = get_secrets_store(str(self.csi_dir))
+        data = secrets_store.get_versioned("secret/versioned-secret")
+        assert data.current == b"current value"
+        assert data.previous == b"previous value"
+
+    def test_simple_encoding(self):
+        secrets_store = get_secrets_store(str(self.csi_dir))
+        data = secrets_store.get_simple("secret/simple-encoded-secret")
+        assert data == b"1337"
+
+    def test_versioned_encoding(self):
+        secrets_store = get_secrets_store(str(self.csi_dir))
+        data = secrets_store.get_versioned("secret/versioned-encoded-secret")
+        assert data.current == b"current encoded value"
+        assert data.previous == b"previous encoded value"
+
+    def test_symlink_updated(self):
+        original_data_path = self.csi_dir.joinpath("..data").resolve()
+        secrets_store = get_secrets_store(str(self.csi_dir))
+        data = secrets_store.get_credentials("secret/bare-secret")
+        assert data.username == "reddit"
+        assert data.password == "password"
+        simulate_secret_update(self.csi_dir)
+        assert original_data_path != self.csi_dir.joinpath("..data").resolve()
+        data = secrets_store.get_credentials("secret/bare-secret")
+        assert data.username == "reddit"
+        assert data.password == "password"
+
+    def test_secret_updated(self):
+        secrets_store = get_secrets_store(str(self.csi_dir))
+        data = secrets_store.get_credentials("secret/example-service/example-secret")
+        gevent.sleep(0.1)  # prevent gevent shenanigans
```

Comment:
What's the purpose of these sleeps?

Reply:
Tests were flaking if they weren't present. My theory is that gevent is passing control back to the tests. By sleeping, I ensure the IO is complete before the remainder of the test.

Reply:
Were the flakes locally or in CI? I've been running the tests a bunch locally and can't seem to reproduce any flakes. I'd like to try to dig into this if possible before we merge.

Reply:
I only saw it in CI

Reply:
Example: https://github.com/reddit/baseplate.py/actions/runs/8116056546/job/22185234411

Reply:
I looked into this a little bit and was able to replicate the issues frequently in GitHub Actions CI but never locally.

I added some debugging prints and noticed cases where the file mtime did not change but the file contents did, which caused the test failures:
<img width="1072" alt="Screenshot 2024-03-07 at 11 53 17 AM" src="https://github.com/reddit/baseplate.py/assets/665269/7809ce5d-0f09-480a-8951-eafbf703fb16">

Not sure if this is something weird with the CI environment (maybe like low-resolution timestamps in the filesystem or with the system clock?) or something with gevent I'm not understanding, but I think it's safe to go ahead and merge.